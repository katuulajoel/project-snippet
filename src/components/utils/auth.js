// eslint-disable-next-line no-unused-vars
import store from "../../store";

export function getAuth() {
  return store.getState().Auth;
}

export function getUser() {
  return getAuth().user;
}

export function isAdmin() {
  let user = getUser();
  return user.is_admin;
}

export function isPayAdmin() {
  let user = getUser();
  return user.is_pay_admin;
}

export function isDev() {
  return getUser().is_developer;
}

export function isDesigner() {
  return getUser().is_designer;
}

export function isClient() {
  return getUser().is_project_owner;
}

export function isPM() {
  return getUser().is_project_manager && !getUser().is_admin;
}

export function isAdminOrClient() {
  return isAdmin() || isClient();
}

export function isAdminOrPM() {
  return isAdmin() || isPM();
}

export function isPayAdminOrPM() {
  return isPayAdmin() || isPM();
}

export function isAdminOrPMOrClient() {
  return isAdmin() || isPM() || isClient();
}

export function isDevOrClient() {
  return isDev() || isClient();
}

export function isDevOrPM() {
  return isDev() || isPM();
}

export function getMyParticipation(project) {
  let myParticipation = null;
  if (project.participation) {
    (project.participation || []).forEach((item) => {
      if (item.user.id === getUser().id) {
        myParticipation = item;
      }
    });
  }
  return myParticipation;
}

export function isPendingProjectParticipant(project) {
  let isPending = false;
  if (isDev()) {
    let myParticipation = getMyParticipation(project);
    if (myParticipation) {
      return myParticipation.status === "initial";
    }
  }
  return isPending;
}

export function hasProjectAccess(project) {
  let allowedUserIds = [];
  ["user", "owner", "pm"].forEach((key) => {
    if (project[key]) {
      allowedUserIds.push(project[key].id);
    }
  });
  if (project.participation) {
    project.participation.forEach((item) => {
      if (item.status === "accepted" && item.user) {
        allowedUserIds.push(item.user.id);
      }
    });
  }
  return isAdmin() || allowedUserIds.includes(getUser().id);
}

export function isPMAndHasProjectAcess(project) {
  return hasProjectAccess(project) && isPM();
}

export function isAdminOrPMWithProjectAcess(project) {
  return isAdmin() || isPMAndHasProjectAcess(project);
}

export function isAdminOrClientOrPMWithProjectAcess(project) {
  return isAdmin() || isClient() || isPMAndHasProjectAcess(project);
}

export function isProjectClient(project) {
  const userId = getUser().id;
  if (project.owner && project.owner.id === userId) {
    return true;
  }

  if (project.user) {
    return project.user.id === userId && (!project.owner || !isAdmin());
  }
  return false;
}
