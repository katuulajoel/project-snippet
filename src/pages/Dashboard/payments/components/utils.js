import {
  isPayAdmin,
  isClient,
  isPayAdminOrPM,
} from "../../../../components/utils/auth";

export const GENERATE_INVOICE_ACTION = "GENERATE_INVOICE_ACTION";
export const PAY_ACTION = "PAY_ACTION";
export const EDIT_ACTION = "EDIT_ACTION";
export const DELETE_ACTION = "DELETE_ACTION";
export const MARK_AS_PAID_ACTION = "MARK_AS_PAID_ACTION";
export const ARCHIVE_ACTION = "ARCHIVE_ACTION";

export const showAction = (action, invoice) => {
  switch (action) {
    case GENERATE_INVOICE_ACTION:
      return (
        isPayAdmin() &&
        !invoice.finalized &&
        !invoice.last_sent_at &&
        !invoice.paid
      );
    case PAY_ACTION:
      return isClient() && invoice.finalized && !invoice.paid;
    case EDIT_ACTION:
    case DELETE_ACTION:
      return (
        isPayAdminOrPM() &&
        !invoice.paid &&
        !invoice.finalized &&
        !invoice.last_sent_at
      );
    case MARK_AS_PAID_ACTION:
      return (
        isPayAdmin() &&
        !invoice.paid &&
        (invoice.finalized || invoice.last_sent_at)
      );
    case ARCHIVE_ACTION:
      return isPayAdmin() && !invoice.paid;
    default:
      return false;
  }
};

// eslint-disable-next-line no-unused-vars
export const performAction = (action, invoice, props) => {
  switch (action) {
    case GENERATE_INVOICE_ACTION: {
      // onGenerateInvoice(invoice.id, props);
      break;
    }
    case PAY_ACTION: {
      /* openModal(<MakePaymentModal invoice={invoice} />, "Make Payment", true, {
        className: "modal-report modal-make-payment",
      }); */
      break;
    }
    case EDIT_ACTION: {
      // onUpdateInvoice(invoice, props);
      break;
    }
    case DELETE_ACTION: {
      // onDeleteInvoice(invoice.id, props);
      break;
    }
    case MARK_AS_PAID_ACTION: {
      // onMarkPaid(invoice.id, props);
      break;
    }
    case ARCHIVE_ACTION: {
      // onMarkArchived(invoice.id, props);
      break;
    }
  }
};
