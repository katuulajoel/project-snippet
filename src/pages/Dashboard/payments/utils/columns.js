// import { isPayAdmin } from "../../../../components/utils/auth";

//TODO: check if use is pay admin to active batch actions, add ...isPayAdmin() && on line 6

export const getTableColumns = (filter) => [
  ...(!(filter === "archived" || filter === "paid")
    ? [
        {
          Header: " ",
          accessor: "batch_action",
        },
      ]
    : []),
  {
    Header: "Date Created",
    accessor: "created_at",
  },
  {
    Header: "Client / Project / Payment Title",
    accessor: "title",
  },
  {
    Header: "Invoice No.",
    accessor: "invoice",
  },
  {
    Header: "Amount",
    accessor: "amount",
  },
  ...(filter !== "paid"
    ? [
        {
          Header: "Due Date",
          accessor: "due_date",
        },
        {
          Header: "Status",
          accessor: "status",
        },
        ...(filter !== "archived"
          ? [
              {
                Header: "",
                accessor: "actions",
              },
            ]
          : []),
      ]
    : []),
];
