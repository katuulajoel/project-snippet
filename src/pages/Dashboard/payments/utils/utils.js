import {
  isPayAdmin,
  isClient,
  isPayAdminOrPM,
} from "../../../../components/utils/auth";
import * as actions from "./constant"; //TODO: Load just actions needed
import {
  approvePayout,
  generateInvoice,
  markAsArchived,
  markAsPaid,
} from "./paymentActions";

function sumInvoiceAmounts(invoices) {
  return invoices
    .map((invoice) => {
      return invoice.amount || 0;
    })
    .reduce((total, number) => {
      return total + Math.round(number * 100) / 100;
    }, 0);
}

/**
 * Group invoices with the same reference into one batch object
 * @param {*} invoices list of invoices
 * @returns returns array of batch objects
 */
export function batchInvoices(invoices) {
  let batchedInvoices = {};
  invoices?.forEach((invoice) => {
    batchedInvoices[invoice.batch_ref] = [
      ...(batchedInvoices[invoice.batch_ref] || []),
      invoice,
    ];
  });

  let batchObjects = [];
  Object.keys(batchedInvoices).map((ref) => {
    let batchObj = {
      ...batchedInvoices[ref][0],
      invoices: batchedInvoices[ref],
    };
    batchObj.ref = ref;
    batchObj.amount = sumInvoiceAmounts(batchedInvoices[ref]);
    batchObjects.push(batchObj);
  });
  return batchObjects;
}

/**
 * Determine whether to show action or not depending on logged in User
 * @param {string} action constants used for the different actions in payments and payouts
 * @param {*} invoice invoice if payments action or invoices if payouts action
 * @returns {boolean} false | true
 */
export const showAction = (action, invoice) => {
  switch (action) {
    case actions.GENERATE_INVOICE_ACTION:
      return (
        isPayAdmin() &&
        !invoice.finalized &&
        !invoice.last_sent_at &&
        !invoice.paid
      );
    case actions.PAY_ACTION:
      return isClient() && invoice.finalized && !invoice.paid;
    case actions.EDIT_ACTION:
    case actions.DELETE_ACTION:
      return (
        isPayAdminOrPM() &&
        !invoice.paid &&
        !invoice.finalized &&
        !invoice.last_sent_at
      );
    case actions.MARK_AS_PAID_ACTION:
      return (
        isPayAdmin() &&
        !invoice.paid &&
        (invoice.finalized || invoice.last_sent_at)
      );
    case actions.APPROVE_BATCH_ACTION:
      return isPayAdmin() && invoice.status !== "approved";
    case actions.ARCHIVE_ACTION:
      return isPayAdmin() && !invoice.paid;
    case actions.EDIT_ACTION_BATCH:
    case actions.DELETE_ACTION_BATCH:
      return (
        isPayAdmin() &&
        !invoice.project.archived &&
        !invoice.status !== "approved" &&
        !invoice.paid
      );
  }
};

/**
 * Determine what function to call depending on the action that been triggered
 * @param {string} action constants used for the different actions in payments and payouts
 * @param {*} data invoice if payments action or invoices if payouts action
 */
export const performAction = (action, data) => {
  switch (action) {
    case actions.GENERATE_INVOICE_ACTION: {
      generateInvoice(data.id);
      break;
    }
    case actions.MARK_AS_PAID_ACTION: {
      markAsPaid(data.id);
      break;
    }
    case actions.ARCHIVE_ACTION: {
      markAsArchived(data.id);
      break;
    }
    case actions.APPROVE_BATCH_ACTION: {
      approvePayout(data);
      break;
    }
  }
};

/**
 * filters to be used as query params
 * @param {*} filter : selected payment status;
 * @returns object keys value pairs
 */
export const getPaymentsFilters = (filter) => {
  switch (filter) {
    case "paid":
      return { paid: "True" };
    case "pending":
      return { paid: "False", overdue: "False" };
    case "overdue":
      return { overdue: "True", paid: "False" };
    case "archived":
      return { archived: "True" };
    default:
      return {};
  }
};
