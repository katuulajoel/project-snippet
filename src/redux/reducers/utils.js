export const getIds = (items) => {
  return Array.from(
    new Set(
      items.map((item) => {
        return item.id;
      })
    )
  );
};

export const reduceUser = (state, user, profile, company) => {
  state = state || {};
  user = user || {};
  profile = { ...state.profile, ...user.profile, ...profile };
  company = { ...state.company, ...user.company, ...company };
  [
    ["name", "company"],
    ["bio", "company_bio"],
    ["website", "website"],
    ["vat_number", "vat_number"],
    ["reg_no", "company_reg_no"],
    ["ref_no", "reference_number"],
    ["skills", "skills"],
    ["country", "country"],
    ["city", "city"],
    ["street", "street"],
    ["plot_number", "plot_number"],
    ["postal_code", "postal_code"],
    ["postal_address", "postal_address"],
    ["tel_number", "phone_number"],
  ].forEach((item) => {
    let companyField = item[0],
      profileField = item[1];
    company[companyField] = company[companyField] || profile[profileField];
  });
  return { ...state, ...user, profile, company };
};
