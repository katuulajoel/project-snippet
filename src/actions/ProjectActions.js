/* eslint-disable no-unused-vars */
import axios from "axios";
import { ENDPOINT_PROJECTS } from "./utils/api";

import {
  FETCH_PROJECT_START,
  FETCH_PROJECT_SUCCESS,
  FETCH_PROJECT_FAILED,
} from "./utils/ActionTypes";
import { success, start, failed } from "./utils/actions";

export function fetchProject(id) {
  return (dispatch) => {
    dispatch(start(FETCH_PROJECT_START));
    axios
      .get(ENDPOINT_PROJECTS + id + "/")
      .then(function (response) {
        dispatch(success(FETCH_PROJECT_SUCCESS, response.data));
      })
      .catch(function (error) {
        dispatch(failed(FETCH_PROJECT_FAILED, error));
      });
  };
}
