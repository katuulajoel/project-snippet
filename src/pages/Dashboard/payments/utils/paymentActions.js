import PropTypes from "prop-types";
import React from "react";
import moment from "moment";
import {
  archiveInvoice,
  generateInvoice as generateInvoiceAction,
  updateInvoice,
  downloadInoicesCsv,
  getInvoiceSummary,
  bulkAction,
} from "../../../../actions/InvoiceActions";
import DateRangeForm from "../../../../components/DateRangeForm";
import ModalHeader from "../../../../components/ModalHeader";
import { openConfirm, openModal } from "../../../../components/utils/modals";
import store from "../../../../store";

const Header = (props) => (
  <ModalHeader
    {...props}
    style={{ paddingBottom: "8px" }}
    options={props.title}
  />
);

Header.propTypes = {
  title: PropTypes.string,
};

export function generateInvoice(invoiceId) {
  openConfirm(
    <div>Are you sure you want to generate an invoice for this payment?</div>,
    null,
    true,
    { ok: "Yes" },
    <Header title="Generate Invoice" />
  ).then(() => {
    store.dispatch(generateInvoiceAction(invoiceId));
  });
}

export function markAsPaid(invoiceId) {
  openConfirm(
    <div>Are you sure you want to mark this invoice as paid?</div>,
    null,
    true,
    { ok: "Yes" },
    <Header title="Mark as Paid" />
  ).then(() => {
    store.dispatch(updateInvoice(invoiceId, { paid: true }));
  });
}

export function markAsArchived(invoiceId) {
  openConfirm(
    <div>Are you sure you want to archive this invoice?</div>,
    null,
    true,
    { ok: "Yes" },
    <Header title="Archive invoice" />
  ).then(() => {
    store.dispatch(archiveInvoice(invoiceId));
  });
}

export function approvePayout(invoices) {
  openConfirm(
    <div>Are you sure you want to approve this payout?</div>,
    null,
    true,
    { ok: "Approve", cancel: "Cancel" },
    <Header title="Approve payout" />
  ).then(() => {
    invoices.forEach((invoice) => {
      store.dispatch(updateInvoice(invoice.id, { status: "approved" }));
    });
  });
}

export function downloadCsv(filter, type) {
  openModal(
    <DateRangeForm id={`export-form`} />,
    "",
    true,
    {
      className: "modal-payments",
      ok: `Download CSV`,
      form: {
        type: "submit",
        form: `export-form`,
      },
    },
    <Header title={`Export ${type}`} />,
    false
  ).then((data) => {
    if (data) {
      let params = {};

      switch (filter) {
        case "archived":
          params["archived"] = "True";
          params["paid"] = "False";
          break;
        case "all":
          params["archived"] = "False";
          break;
        case "pending":
          params["archived"] = "False";
          params["paid"] = "False";
          params["overdue"] = "False";
          break;
        case "overdue":
          params["archived"] = "False";
          params["paid"] = "False";
          params["overdue"] = "True";
          break;
        case "paid":
          params["archived"] = "False";
          params["paid"] = "True";
      }

      store.dispatch(
        downloadInoicesCsv({
          start: data.start,
          end: data.end,
          type: type === "Payments" ? "sale" : "purchase",
          ...params,
        })
      );
    }
  });
}

export function filterPayment({ start, end }, type, setSummariesRange) {
  openModal(
    <DateRangeForm
      id={`fiter-payment`}
      defaultStart={start}
      defaultEnd={end}
      maxdate={new Date()}
      message={"Select date range for filter"}
    />,
    "",
    true,
    {
      className: "modal-payments",
      ok: `Submit`,
      form: {
        type: "submit",
        form: `fiter-payment`,
      },
    },
    <ModalHeader
      style={{ paddingBottom: "8px" }}
      options={{
        title: `Filter Total ${type === "Payments" ? "Payments" : "Payouts"}`,
      }}
    />,
    false
  ).then((data) => {
    if (data) {
      const START = `${moment(data.start).format(
        moment.HTML5_FMT.DATE
      )}T00:00:00`;
      const END = `${moment(data.end).format(moment.HTML5_FMT.DATE)}T23:59:59`;

      setSummariesRange({
        start: START,
        end: END,
      });

      store.dispatch(
        getInvoiceSummary({
          min_date: START,
          max_date: END,
          type: type === "Payments" ? "sale" : "purchase",
        })
      );
    }
  });
}

export function bulkMarkAsPaid(invoices) {
  openConfirm(
    <div>Are you sure you want to mark all selected as paid?</div>,
    null,
    true,
    { ok: "Yes", cancel: "Cancel" },
    <Header title="Bulk mark as Paid" />
  ).then(() => {
    invoices.forEach((invoice) => {
      store.dispatch(updateInvoice(invoice.id, { paid: true }));
    });
  });
}

export function bulkGenerateInvoice(invoices) {
  openConfirm(
    <div>Are you sure you want to generate invoices for all selected?</div>,
    null,
    true,
    { ok: "Yes", cancel: "Cancel" },
    <Header title="Bulk generate invoices" />
  ).then(() => {
    invoices.forEach((invoice) => {
      store.dispatch(generateInvoice(invoice.id, { paid: true }));
    });
  });
}

export function bulkDeleteInvoice(invoices) {
  openConfirm(
    <div>Are you sure you want to delete all selected?</div>,
    null,
    true,
    { ok: "Yes", cancel: "Cancel" },
    <Header title="Bulk delete invoices" />
  ).then(() => {
    store.dispatch(bulkAction(invoices, "archive"));
  });
}

export function bulkArchiveInvoice(invoices) {
  openConfirm(
    <div>Are you sure you want to archive all selected?</div>,
    null,
    true,
    { ok: "Yes", cancel: "Cancel" },
    <Header title="Bulk archive invoices" />
  ).then(() => {
    store.dispatch(bulkAction(invoices, "delete"));
  });
}
