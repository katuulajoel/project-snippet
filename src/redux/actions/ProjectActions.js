/* eslint-disable no-unused-vars */
import axios from "axios";
import {
  composeFormData,
  ENDPOINT_PROJECTS,
  ENDPOINT_PROGRESS_EVENTS,
  ENDPOINT_DOCUMENTS,
  ENDPOINT_ACTIVITIES,
} from "../../utils/api";

import {
  FETCH_PROJECT_START,
  FETCH_PROJECT_SUCCESS,
  FETCH_PROJECT_FAILED,
  FETCH_PROJECTS_START,
  FETCH_PROJECTS_SUCCESS,
  FETCH_PROJECTS_FAILED,
  FETCH_MORE_PROJECTS_START,
  FETCH_MORE_PROJECTS_SUCCESS,
  FETCH_MORE_PROJECTS_FAILED,
  UPDATE_PROJECT_START,
  UPDATE_PROJECT_SUCCESS,
  UPDATE_PROJECT_FAILED,
  CREATE_PROGRESS_EVENT_START,
  CREATE_PROGRESS_EVENT_SUCCESS,
  CREATE_PROGRESS_EVENT_FAILED,
  UPDATE_PROGRESS_EVENT_START,
  UPDATE_PROGRESS_EVENT_SUCCESS,
  UPDATE_PROGRESS_EVENT_FAILED,
  CREATE_DOCUMENT_START,
  CREATE_DOCUMENT_SUCCESS,
  CREATE_DOCUMENT_FAILED,
  UPDATE_DOCUMENT_START,
  UPDATE_DOCUMENT_SUCCESS,
  UPDATE_DOCUMENT_FAILED,
  DELETE_DOCUMENT_START,
  DELETE_DOCUMENT_SUCCESS,
  DELETE_DOCUMENT_FAILED,
  LIST_TIMESHEETS_START,
  LIST_TIMESHEETS_SUCCESS,
  LIST_TIMESHEETS_FAILED,
  CREATE_TIMESHEET_START,
  CREATE_TIMESHEET_SUCCESS,
  CREATE_TIMESHEET_FAILED,
  LIST_ACTIVITIES_START,
  LIST_ACTIVITIES_SUCCESS,
  LIST_ACTIVITIES_FAILED,
  LIST_MORE_ACTIVITIES_START,
  LIST_MORE_ACTIVITIES_SUCCESS,
  LIST_MORE_ACTIVITIES_FAILED,
} from "../../configs/constants/ActionTypes";
import { success, start, failed } from "../../utils/actions";

export function fetchProject(id) {
  return (dispatch) => {
    dispatch(start(FETCH_PROJECT_START));
    axios
      .get(ENDPOINT_PROJECTS + id + "/")
      .then(function (response) {
        dispatch(success(FETCH_PROJECT_SUCCESS, response.data));
      })
      .catch(function (error) {
        dispatch(failed(FETCH_PROJECT_FAILED, error));
      });
  };
}

export function fetchProjects(filter) {
  return (dispatch) => {
    dispatch(start(FETCH_PROJECTS_START));
    axios
      .get(
        Object.prototype.hasOwnProperty.call(filter, "archived_type")
          ? ENDPOINT_PROJECTS + "archived/"
          : ENDPOINT_PROJECTS,
        { params: filter }
      )
      .then(function (response) {
        dispatch(success(FETCH_PROJECTS_SUCCESS, response.data));
      })
      .catch(function (error) {
        dispatch(failed(FETCH_PROJECTS_FAILED, error));
      });
  };
}

export function fetchMoreProjects(nextUrl) {
  return (dispatch) => {
    dispatch(start(FETCH_MORE_PROJECTS_START));
    axios
      .get(nextUrl)
      .then(function (response) {
        dispatch(success(FETCH_MORE_PROJECTS_SUCCESS, response.data));
      })
      .catch(function (error) {
        dispatch(failed(FETCH_MORE_PROJECTS_FAILED, error));
      });
  };
}

export function updateProject(id, project) {
  return (dispatch) => {
    dispatch(start(UPDATE_PROJECT_START));

    let headers = {},
      data = project;
    if (project.documents && project.documents.length) {
      headers["Content-Type"] = "multipart/form-data";
      data = composeFormData(project);
    }

    axios
      .patch(`${ENDPOINT_PROJECTS}${id}/`, data, {
        headers: { ...headers },
      })
      .then(function (response) {
        dispatch(success(UPDATE_PROJECT_SUCCESS, response.data));
      })
      .catch(function (error) {
        dispatch(failed(UPDATE_PROJECT_FAILED, error));
      });
  };
}

export function createProgressEvent(progress_event) {
  return (dispatch) => {
    dispatch(start(CREATE_PROGRESS_EVENT_START));
    axios
      .post(ENDPOINT_PROGRESS_EVENTS, progress_event)
      .then(function (response) {
        dispatch(success(CREATE_PROGRESS_EVENT_SUCCESS, response.data));
      })
      .catch(function (error) {
        dispatch(failed(CREATE_PROGRESS_EVENT_FAILED, error));
      });
  };
}

export function updateProgressEvent(id, progress_event) {
  return (dispatch) => {
    dispatch(start(UPDATE_PROGRESS_EVENT_START));
    axios
      .patch(ENDPOINT_PROGRESS_EVENTS + id + "/", progress_event)
      .then(function (response) {
        dispatch(success(UPDATE_PROGRESS_EVENT_SUCCESS, response.data));
      })
      .catch(function (error) {
        dispatch(failed(UPDATE_PROGRESS_EVENT_FAILED, error));
      });
  };
}

export function createDocument(document) {
  return (dispatch) => {
    dispatch(start(CREATE_DOCUMENT_START));

    let headers = {},
      data = document;

    if (document.file) {
      headers["Content-Type"] = "multipart/form-data";
      data = composeFormData(document);
    }

    axios
      .post(ENDPOINT_DOCUMENTS, data, { headers })
      .then(function (response) {
        dispatch(success(CREATE_DOCUMENT_SUCCESS, response.data));
      })
      .catch(function (error) {
        dispatch(failed(CREATE_DOCUMENT_FAILED, error));
      });
  };
}

export function updateDocument(id, document) {
  return (dispatch) => {
    dispatch(start(UPDATE_DOCUMENT_START));

    let headers = {},
      data = document;
    if (document.file) {
      headers["Content-Type"] = "multipart/form-data";
      data = composeFormData(document);
    }

    axios
      .patch(ENDPOINT_DOCUMENTS + id + "/", data, {
        headers: { ...headers },
      })
      .then(function (response) {
        dispatch(success(UPDATE_DOCUMENT_SUCCESS, response.data));
      })
      .catch(function (error) {
        dispatch(failed(UPDATE_DOCUMENT_FAILED, error));
      });
  };
}

export function deleteDocument(id) {
  return (dispatch) => {
    dispatch(start(DELETE_DOCUMENT_START));
    axios
      .delete(ENDPOINT_DOCUMENTS + id + "/")
      .then(function () {
        dispatch(success(DELETE_DOCUMENT_SUCCESS, id));
      })
      .catch(function (error) {
        dispatch(failed(DELETE_DOCUMENT_FAILED, error));
      });
  };
}

export function listTimesheets(id, year) {
  return (dispatch) => {
    dispatch(start(LIST_TIMESHEETS_START));
    axios
      .get(`${ENDPOINT_PROJECTS}${id}/timesheets/?year=${year}`)
      .then((response) => {
        dispatch(success(LIST_TIMESHEETS_SUCCESS, response.data));
      })
      .catch((error) => {
        dispatch(failed(LIST_TIMESHEETS_FAILED, error));
      });
  };
}

export function createTimesheet(timesheet, id) {
  return (dispatch) => {
    dispatch(start(CREATE_TIMESHEET_START));
    axios
      .post(`${ENDPOINT_PROJECTS}${id}/timesheet/`, timesheet)
      .then(function (response) {
        dispatch(success(CREATE_TIMESHEET_SUCCESS, response.data));
      })
      .catch(function (error) {
        dispatch(failed(CREATE_TIMESHEET_FAILED, error));
      });
  };
}

export function listActivities(filter) {
  return (dispatch) => {
    dispatch(start(LIST_ACTIVITIES_START));
    axios
      .get(ENDPOINT_ACTIVITIES, { params: filter })
      .then(function (response) {
        dispatch(success(LIST_ACTIVITIES_SUCCESS, response.data));
      })
      .catch(function (error) {
        dispatch(failed(LIST_ACTIVITIES_FAILED, error));
      });
  };
}

export function listMoreActivities(url) {
  return (dispatch) => {
    dispatch(start(LIST_MORE_ACTIVITIES_START));
    axios
      .get(url)
      .then(function (response) {
        dispatch(success(LIST_MORE_ACTIVITIES_SUCCESS, response.data));
      })
      .catch(function (error) {
        dispatch(failed(LIST_MORE_ACTIVITIES_FAILED, error));
      });
  };
}
