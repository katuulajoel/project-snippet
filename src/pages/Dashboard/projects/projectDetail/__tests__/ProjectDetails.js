/* eslint-disable no-unused-vars */ // TODO:
import React from "react";
import renderer from "react-test-renderer";
import ProjectPage from "..";
import { BrowserRouter } from "react-router-dom";
import { Provider } from "react-redux";
import configureStore from "redux-mock-store";
import thunk from "redux-thunk";
import { ThemeProvider } from "styled-components";
import theme from "../../../../../assets/theme";
import * as actions from "../../../../../redux/actions/ProjectActions";
import * as access from "../../../../../utils/auth";

const middlewares = [thunk];

const mockAppState = {
  Invoice: {
    isMakingRequest: {},
    errors: {},
    summary: {},
    list: { data: [], count: 0, next: "", previous: "" },
    invoice: {},
    csv: {},
  },
  Projects: {
    isMakingRequest: {},
    project: { id: 124 },
  },
};

const mockAppStore = (state) => {
  const mockStore = configureStore(middlewares);
  return mockStore(state);
};

describe("Dashboard test", () => {
  let fetchProjectStub = "";
  let hasProjectAccessStub = "";
  beforeAll(() => {
    fetchProjectStub = jest.spyOn(actions, "fetchProject");
    hasProjectAccessStub = jest
      .spyOn(access, "hasProjectAccess")
      .mockReturnValue(true);
  });
  it("Should match snapshot test", () => {
    global.URL.createObjectURL = jest.fn(() => "details");
    const tree = renderer
      .create(
        <BrowserRouter>
          <Provider store={mockAppStore(mockAppState)}>
            <ThemeProvider theme={theme}>
              <ProjectPage match={{ path: "/projects", params: { id: 123 } }} />
            </ThemeProvider>
          </Provider>
        </BrowserRouter>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("should show loading indicator when no project or requesting project", () => {
    const tree = renderer
      .create(
        <BrowserRouter>
          <Provider
            store={mockAppStore({
              ...mockAppState,
              Projects: { isMakingRequest: {}, project: null },
            })}
          >
            <ThemeProvider theme={theme}>
              <ProjectPage match={{ path: "/projects" }} />
            </ThemeProvider>
          </Provider>
        </BrowserRouter>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
