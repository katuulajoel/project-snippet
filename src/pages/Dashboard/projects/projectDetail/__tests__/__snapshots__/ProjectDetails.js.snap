// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dashboard test Should match snapshot test 1`] = `null`;

exports[`Dashboard test should show loading indicator when no project or requesting project 1`] = `
<div
  className="loading"
  data-testid="loading"
  style={
    Object {
      "textAlign": "center",
    }
  }
>
  <img
    src={Object {}}
    style={
      Object {
        "width": "20px",
      }
    }
  />
   
  
</div>
`;
